feat: new module for fullaudio (microphone) communication

mariogasparoni opened the pull request:
Adds a new module (FullAudio) which is able to handle microphone requests (send/receive audio).
This new module routes microphone flow from Browser <-> MediaServer (kurento or mediasoup) <-> FreeSWITCH.
This is still experimental and currently works with Kurento only (though i already tested a different code with success using mediasoup).
I will keep working on this to fully enable two-way audio using mediasoup and at some point we will switch to it as default media server for both microphone and listenonly.

mariogasparoni committed:
feat(audio): Add base files for fullaudio module

This module will deal with microphone negotiation

mariogasparoni committed:
feat(audio): allow sendrecv (bbb's microphone) stream in fullaudio module

The stream is created using Kurento, but we are going to update code
to work for mediasoup also.
Added config to switch fullaudio's adapter (currently working with Kurento's
only)

mariogasparoni committed:
fix(audio): correctly call fullaudio module from it's process and manager

mariogasparoni committed:
fix(audio): correctly removes user from audio conference in fullaudio adapter

We now create a new session for each connection made to bbb-webrtc-sfu, and
only remove the connection that belongs to the user who is leaving audio.
Sessions are now stored by user's connectionId

mariogasparoni committed:
fix(audio): do not try to set conference floor when session already ended

Log now doesn't show error when this happens

mariogasparoni committed:
fix(audio): fix delay to start processing mic stream.

Disabled comedia for fullaudio bridge.

mariogasparoni committed:
fix(fullaudio): correctly disconnect all users from session

This function is called with end-meeting and needs to iterate over all
sessions, since we now work with a session per user

mariogasparoni committed:
fix(fullaudio): do not log USER_NOT_FOUND error when user already left the room

After leaving fullaudio, the ejection check were triggering this error log
unnecessarily.

prlanzarin commented on config/default.example.yml:
```diff
@@ -48,6 +48,7 @@ videoSubscriberSpecSlave: false
 videoMediaServer: Kurento
 screenshareMediaServer: Kurento
 audioMediaServer: Kurento
+fullAudioMediaServer: Kurento
```
Please add an environment variable equivalent to custom-environment-variables.yml

mariogasparoni replied:
Sure, will do it

mariogasparoni replied:
updated PR

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+    }
+  }
+
+  async _handleUserJoinedVoiceConf (payload) {
```
This method can be removed; it's enough that it is in the listen only module already.

mariogasparoni replied:
Created an issue for this: https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/107

prlanzarin commented on lib/audio/fullaudio.js:
```diff
@@ -0,0 +1,1014 @@
+    this.sourceAudioStatus = C.MEDIA_STOPPED;
+  }
+
+  sendUserDisconnectedFromGlobalAudioMessage(connectionId) {
```
This method can be removed (listen only specific)

mariogasparoni replied:
Created an issue for this: https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/107

prlanzarin commented on lib/audio/fullaudio.js:
```diff
@@ -0,0 +1,1014 @@
+    }
+  };
+
+  sendUserConnectedToGlobalAudioMessage(connectionId) {
```
This method can be removed (listen only specific)

mariogasparoni replied:
Created an issue for this: https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/107

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+const ERRORS = require('../base/errors.js');
+
+const { handleExternalConnections : FS_HANDLE_EXTERNAL_CONNECTIONS }
+  = config.get('freeswitch');
```
```suggestion
```

mariogasparoni replied:
updated PR

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+      const meetingId = payload[C.MEETING_ID_2x];
+      this._disconnectAllUsers(meetingId);
+    });
+    this._bbbGW.on(C.USER_JOINED_VOICE_CONF_MESSAGE_2x, this._handleUserJoinedVoiceConf.bind(this));
```
```suggestion
    this._bbbGW.on(C.USER_JOINED_VOICE_CONF_MESSAGE_2x, this._handleUserJoinedVoiceConf.bind(this));
```

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+    if (session) {
+      const connectionId = session.getConnectionId(userId);
+      if (connectionId) {
+        Logger.info(this._logPrefix, 'Disconnect listen only session on UserLeft*', {
```
```suggestion
        Logger.info(this._logPrefix, 'Disconnect full audio session on UserLeft*', {
```

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+          sdpAnswer : sdpAnswer
+        }, C.FROM_AUDIO);
+
+        Logger.info(this._logPrefix, `Started listen only session for user ${userId}`,
```
```suggestion
        Logger.info(this._logPrefix, `Started full audio session for user ${userId}`,
```

prlanzarin commented on lib/audio/fullaudio.js:
```diff
@@ -0,0 +1,1014 @@
+      connectionId: connectionId,
+      id: "webRTCAudioSuccess",
+      success: "MEDIA_FLOWING"
+    }, C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

prlanzarin commented on lib/audio/fullaudio.js:
```diff
@@ -0,0 +1,1014 @@
+      connectionId: connectionId,
+      id: "webRTCAudioError",
+      error: { code: 2211 , reason: errors[2211] },
+    }, C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

prlanzarin commented on lib/audio/fullaudio.js:
```diff
@@ -0,0 +1,1014 @@
+      connectionId: connectionId,
+      id: "webRTCAudioError",
+      error: { code: 2211 , reason: errors[2211] },
+    }, C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

prlanzarin commented on lib/audio/fullaudio.js:
```diff
@@ -0,0 +1,1014 @@
+      connectionId,
+      id : 'iceCandidate',
+      candidate : candidate
+    }, C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+            connectionId,
+            type: C.AUDIO_APP,
+            id : 'close',
+          }), C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+    this.sendToClient({
+      type: 'audio',
+      ...error,
+    }, C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+    this.sendToClient(({
+      type: 'audio',
+      ...error,
+    }), C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

mariogasparoni replied:
Created an issue for this: https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/108

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+          id : 'startResponse',
+          response : 'accepted',
+          sdpAnswer : sdpAnswer
+        }, C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

mariogasparoni replied:
Created an issue for this: https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/108

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+        this.sendToClient({
+          type: 'audio',
+          ...errorMessage,
+        }, C.FROM_AUDIO);
```
Current IPC is not via Redis so this is mostly harmless, but I suggest creating a new channel constant (`C.FROM_FULL_AUDIO/TO_FULL_AUDIO) to avoid unexpected problems in the future.

mariogasparoni replied:
Created an issue for this: https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/108

prlanzarin commented on lib/mcs-core/lib/media/media-controller.js:
```diff
@@ -145,9 +152,18 @@ module.exports = class MediaController {
+
       // User or room were already closed or not found, resolving as it is
       const normalizedError = this._handleError(error);
-      Logger.warn(LOG_PREFIX, `Leave for ${userId} failed due to ${error.message}`, { roomId, userId, error });
```
Please keep this warning. It's there as a reminder that the leave procedure is botched by design and should not fail that easily, hence that it needs to be fixed.

mariogasparoni replied:
Sure

prlanzarin commented on lib/audio/FullAudioManager.js:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  _handleSessionWideError (error, sessionId, rawMessage) {
+    Logger.error(this._logPrefix, `Listen only session wide fatal failure`, {
```
```suggestion
    Logger.error(this._logPrefix, `Full audio session wide fatal failure`, {
```

mariogasparoni replied:
updated PR

prlanzarin reviewed (commented):
Very early cursory review. There's more stuff to be discussed.

prlanzarin commented:
As a general review: this new module seems to double as a full audio + listen only module.

My suggestion is to get rid of all listen only-related code that was ported from the listen only module and keep this bridge bidirectional only. Making it leaner will make it easier to extend it or even give you more freedom on doing stuff differently than it is currently. Listen only should be left to the module that does it already.

When we go forward with removing mic vs listen only separation in the client, I see it as a mostly backend-only change.
The client should always see it's audio/voice stream as a bidirectional one, and it should always think it's connected to microphone. The unification will be done server side, by us, based on mute state, so the only listen only logic that will have to be _ported_ over is the GLOBAL_AUDIO bridge maintenance. But, again, I'd prefer if we started lean and ported (or re-implemented) that when necessary.

mariogasparoni commented:
Thanks @prlanzarin. 
Yes this module is basically a copy of the current audio module (listenonly) + additions for microphone.
My first idea was to keep listenonly working in this module without regressions, while we could also add support for fullaudio (microphone). I am fine with removing listenonly code from this, and trying to keep things transparent.

Since, this doesn't break current listenonly, i created an issue for this: https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/107
so we can work on this in future versions.

About all other changes, i will be working on it.

mariogasparoni committed:
update(fullaudio): add FULL_AUDIO_MEDIA_SERVER environment variable

This sets the value for fullAudioMediaServer setting

mariogasparoni committed:
update(fullaudio): keep the USER_NOT_FOUND warning  when calling leave()

Update according to @prlanzarin 's comment here:
https://github.com/bigbluebutton/bbb-webrtc-sfu/pull/106#discussion_r733177686

prlanzarin committed:
update(fullaudio): fix log error string according to fullaudio module

prlanzarin committed:
update(fullaudio): remove handleExternalConnections from fullaudio manager

Updating according to @prlanzarin 's suggestion:
https://github.com/bigbluebutton/bbb-webrtc-sfu/pull/106#discussion_r733169810

mariogasparoni committed:
update(fullaudio): remove reference to undefined variable
